e2e/pull/pull.go

```diff
@@ -145,79 +150,90 @@ var tests = []struct {
-		force:         true,
-		expectSuccess: false,
+		desc:             "oras pull of non SIF file",
+		srcURI:           "oras://localhost:5000/pull_test_:latest", // TODO(mem): obtain registry from context
+		force:            true,
+		expectedExitCode: 255,
 	},
 	{
-		desc:          "oras pull of packed dir",
-		srcURI:        "oras://localhost:5000/pull_test_invalid_file:latest", // TODO(mem): obtain registry from context
-		force:         true,
-		expectSuccess: false,
+		desc:             "oras pull of packed dir",
+		srcURI:           "oras://localhost:5000/pull_test_invalid_file:latest", // TODO(mem): obtain registry from context
+		force:            true,
+		expectedExitCode: 255,
 	},
 
 	// pulling with library URI argument
 	{
-		desc:          "bad library URI",
-		srcURI:        "library://busybox",
-		library:       "https://bad-library.sylabs.io",
-		expectSuccess: false,
+		desc:             "bad library URI",
+		srcURI:           "library://busybox",
+		library:          "https://bad-library.sylabs.io",
+		expectedExitCode: 255,
 	},
 	{
-		desc:          "default library URI",
-		srcURI:        "library://busybox",
-		library:       "https://library.sylabs.io",
-		force:         true,
-		expectSuccess: true,
+		desc:             "default library URI",
+		srcURI:           "library://busybox",
+		library:          "https://library.sylabs.io",
+		force:            true,
+		expectedExitCode: 0,
 	},
 }
 
-func (c *ctx) imagePull(t *testing.T, imgURI, library, pullDir, imagePath string, force, unauthenticated bool) (string, []byte, error) {
-	argv := []string{"pull"}
+func (c *ctx) imagePull(t *testing.T, tt testStruct) {
+	// We use a string rather than a slice of strings to avoid having an empty
+	// element in the slice, which would cause the command to fail, wihtout
+	// over-complicating the code.
+	argv := ""
 
-	if force {
-		argv = append(argv, "--force")
+	if tt.force {
+		argv += "--force "
 	}
 
-	if unauthenticated {
-		argv = append(argv, "--allow-unauthenticated")
+	if tt.unauthenticated {
+		argv += "--allow-unauthenticated "
 	}
 
-	if pullDir != "" {
-		argv = append(argv, "--dir", pullDir)
+	if tt.pullDir != "" {
+		argv += "--dir " + tt.pullDir + " "
 	}
 
-	if library != "" {
-		argv = append(argv, "--library", library)
+	if tt.library != "" {
+		argv += "--library " + tt.library + " "
 	}
 
-	if imagePath != "" {
-		argv = append(argv, imagePath)
+	if tt.imagePath != "" {
+		argv += tt.imagePath + " "
 	}
 
-	argv = append(argv, imgURI)
-
-	cmd := fmt.Sprintf("%s %s", c.env.CmdPath, strings.Join(argv, " "))
-	out, err := exec.Command(c.env.CmdPath, argv...).CombinedOutput()
-
-	return cmd, out, err
+	argv += tt.srcURI
+
+	e2e.RunSingularity(
+		t,
+		tt.desc,
+		e2e.WithPrivileges(false),
+		e2e.WithCommand("pull"),
+		e2e.WithArgs(strings.Split(argv, " ")...),
+		e2e.WithEnv(append(os.Environ(), "c.env.KeyringDir")), // We make sure to include Environ to find utilities such as mksquashfs
```

Comment:
if you doesn't pass `WithEnv`, `RunSingularity` takes `os.Environ` by default

Reply:
I need pass `c.env.KeyringDir` so based on my testing, i need to pass os.Environ *and* `c.env.KeyringDir`. I will update the comment.